README.md
# vscode_irite

CHANGELOG.md
ChangeLog
=========

#### 2.12.1
- Security updates. 

#### 2.12.0
- Security updates.  882D Snippets and highlighting added.  

#### 2.11.0
- SubmitDSPData snippet actually filled in SubmitData code, fixed.  Missing snippet for GetFBStatus.  

#### 2.10.1

- Added new API for getting the status of all DIOs on a given slot with one bitmasked integer.

#### 2.9.0

- Missing highlighting from recent changes

#### 2.8.0

- Added new snippets and highlighting for the new APIs

#### 2.7.1

- Security patches

#### 2.7.0

- Completely revamped the snippets for continuity and completeness.
- Added missing or incorrect snippet and autocomplete info.

#### 2.5.2

-   Update 1.33 to VSCode broke the extension. Can no longer compile the
    active window. No longer will try to search for .src just compile
    first one. Will work like used to now.  MUST HAVE .src file ACTIVE to
    compile.

#### 2.5.1

-   Fixed issue when multiple .src files are present in the
    folder/workspace only the first one alphabetically would compile.
    Now if only one exists will find and compile that one regardless of
    what file is active. If more than one src file is found in the
    folder it will compile the active window. User needs to make sure
    they have the correct .src file active when compiling.

#### 2.4.0

-   Fixed issue in which the console did not auto-scroll to the bottom
    so any compiler success/fail message as well as any compiler issues
    did not show.

#### 2.3.1

-   Fixed issue in which the compiler success/fail message as well as
    any compiler issues did not show in the console.

#### 2.2.0

-   Patched security flaw introduced by flatmap-stream malware.

#### 2.1.0

-   Fixed error when trying to build with only a file open.
    Workspace/folder is required.

#### 2.0.2

-   Autosaves all workspace files and selects the src before building.
    -   This allows autosaving to be turned off and initiating the build
        from any file AND any window.
        -   (No more trying to compile the output window or iri file)

#### 2.0.1

-   Added missing syntax colorization and snippets
-   Added more detailed description to each snippet
-   Added and fixed the tab to complete portions of each snippet
-   Complete revamp and expansion of Intellisense.



src/build.ts
import * as vscode from 'vscode';
import fs = require('fs');
import cp = require('child_process');
import os = require('os');
import npath = require('path');

export class Builder {

    constructor(private _document: vscode.TextDocument, private _options?: vscode.FormattingOptions) {
        this._options = this._options || {
            insertSpaces: false,
            tabSize: 2
        };

        if (typeof this._options.insertSpaces === 'undefined') {
            this._options.insertSpaces = false;
            this._options.tabSize = 2;
        }

        if (typeof this._options.tabSize != 'number' || isNaN(this._options.tabSize)) {
            this._options.tabSize = 2;
        }

        this._options.tabSize = Math.max(0, 2);
    }

    public format(range: vscode.Range, engine: string, path: string, parameters: string, indent: number, wrapLineLength: number) {

        return new Promise((resolve, reject) => {

            // entire document - if not range is provided
            range = range || new vscode.Range(
                0, 0,
                this._document.lineCount,
                this._document.lineAt(this._document.lineCount - 1).range.end.character
            );

            let textToFormat = this._document.getText(range);
            let tempFile: string = npath.join(os.tmpdir(), 'tmp.tmp.pas');
            let command: string;
            let tempFileOut: string = npath.join(os.tmpdir(), 'tmp.tmp.out');
            let configFileParameters: string = '';

            fs.writeFileSync(tempFile, textToFormat);

            if (textToFormat) {

                try {

                    //
                    if (engine == 'ptop') {
                        if (parameters != '') {
                            configFileParameters = ' -c ' + parameters;
                        }

                        let indentConfig: string = '';
                        if (indent > 0) {
                            indentConfig = ' -i ' + indent
                        }

                        let wrapLineLengthConfig: string = '';
                        if (wrapLineLength > 0) {
                            wrapLineLengthConfig = ' -l ' + wrapLineLength
                        }

                        command = "\"" + path + "\" " + configFileParameters + indentConfig + wrapLineLengthConfig + ' \"$file\" \"$outfile\" ';
                        command = command.replace('$file', tempFile);
                        command = command.replace('$outfile', tempFileOut);
                    } else { // jcf
                        if (parameters != '') {
                            configFileParameters = ' -config=' + parameters;
                        }
                        command = "\"" + path + "\" " + configFileParameters + '  -y -F \"$file\" ';
                        command = command.replace('$file', tempFile);
                    }

                    console.log(command);
                    cp.exec(command, function(error, stdout, stderr) {
                        if (error) {
                            reject(stdout.toString());
                        }
                        else {
                            let formattedXml: string = fs.readFileSync(tempFileOut, 'utf8');
                            resolve(formattedXml);;
                        }
                    });

                } catch (err) {
                    reject(err.toString());
                }

            }
            // the parser didn't return anything we can use, show an error message and return
            else {
                reject('no text to format');
            }
        });

    }

}

src/comparers.ts
'use strict';
import { TextDocument, TextEditor, Uri } from 'vscode';

abstract class Comparer<T> {
    abstract equals(lhs: T, rhs: T): boolean;
}

class UriComparer extends Comparer<Uri> {

    equals(lhs: Uri | undefined, rhs: Uri | undefined) {
        if (lhs === undefined && rhs === undefined) return true;
        if (lhs === undefined || rhs === undefined) return false;

        return lhs.scheme === rhs.scheme && lhs.fsPath === rhs.fsPath;
    }
}

class TextDocumentComparer extends Comparer<TextDocument> {

    equals(lhs: TextDocument | undefined, rhs: TextDocument | undefined) {
        if (lhs === undefined && rhs === undefined) return true;
        if (lhs === undefined || rhs === undefined) return false;

        return uriComparer.equals(lhs.uri, rhs.uri);
    }
}

class TextEditorComparer extends Comparer<TextEditor> {

    equals(lhs: TextEditor | undefined, rhs: TextEditor | undefined, options: { useId: boolean, usePosition: boolean } = { useId: false, usePosition: false }) {
        if (lhs === undefined && rhs === undefined) return true;
        if (lhs === undefined || rhs === undefined) return false;

        if (options.usePosition && (lhs.viewColumn !== rhs.viewColumn)) return false;

        if (options.useId && (!lhs.document || !rhs.document)) {
            if ((lhs as any)._id !== (rhs as any)._id) return false;

            return true;
        }

        return textDocumentComparer.equals(lhs.document, rhs.document);
    }
}

const textDocumentComparer = new TextDocumentComparer();
const textEditorComparer = new TextEditorComparer();
const uriComparer = new UriComparer();
export {
    textDocumentComparer as TextDocumentComparer,
    textEditorComparer as TextEditorComparer,
    uriComparer as UriComparer
};

webpack.config.js
//@ts-check

'use strict';

const path = require('path');

/**@type {import('webpack').Configuration}*/
const config = {
    target: 'node', // vscode extensions run in a Node.js-context 📖 -> https://webpack.js.org/configuration/node/

    entry: './src/extension.ts', // the entry point of this extension, 📖 -> https://webpack.js.org/configuration/entry-context/
    output: {
        // the bundle is stored in the 'dist' folder (check package.json), 📖 -> https://webpack.js.org/configuration/output/
        path: path.resolve(__dirname, 'dist'),
        filename: 'extension.js',
        libraryTarget: 'commonjs2',
        devtoolModuleFilenameTemplate: '../[resource-path]'
    },
    devtool: 'source-map',
    externals: {
        vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
    },
    resolve: {
        // support reading TypeScript and JavaScript files, 📖 -> https://github.com/TypeStrong/ts-loader
        extensions: [ '.ts', '.js' ]
    },
    module: {
        rules: [
            {
                test: /\.ts$/,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'ts-loader'
                    }
                ]
            }
        ]
    }
};
module.exports = config;

src/extension - Copy.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";
//import * as builder from './build'; //removed until language server build starts
import fs = require("fs");
import path = require("path");
import cp = require("child_process");
var opener = require("opener");
var iRiteChannel = vscode.window.createOutputChannel("iRite Information");

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  interface EngineParams {
    engine: string;
    enginePath: string;
  }

  // Use the console to output diagnostic information (console.log) and errors (console.error)
  // This line of code will only be executed once when your extension is activated
  console.log("Congratulations, iRite is now active!");

  //check to see if the .exe is in specified path, if not display message to download revolution
  let enginePath: string = vscode.workspace
    .getConfiguration("irite")
    .get("build.enginePath", "");
  var optionRev = <vscode.MessageItem>{ title: "Get Revolution" };

  fs.stat(enginePath, function(err, stat) {
    if (err == null) {
      console.log("Revolution exists");
    } else if (err.code == "ENOENT") {
      // file does not exist
      vscode.window
        .showInformationMessage(
          "Revolution can not be located, would you like to download it?",
          optionRev
        )
        .then(option => {
          if (typeof option == "undefined") {
            return;
          }
          switch (option.title) {
            case optionRev.title:
              opener(
                "https://www.ricelake.com/en-us/products/product-details/revolution-scale-software#/resources-downloads"
              );
              break;
            default:
              break;
          }
        });
    } else {
      console.log("Revolution Checker: ", err.code);
    }
  });

  //registercommands takes reference from package.json, alter that file for more/different commands
  vscode.commands.registerCommand("irite.build", () => {
    iRiteChannel.clear();
    vscode.workspace.saveAll().then(
      success => {
        if (success) {
          iRiteChannel.appendLine("All Files Saved");
          var openTextDoc;
          iRiteChannel.appendLine("Result: " + result.toString());
          if (result.length > 0) {
            {
              // if(result.length > 1){
                openTextDoc = vscode.window.activeTextEditor.document;

              // } else {
              //   openTextDoc = vscode.workspace.openTextDocument(result[0]);
              // }

              iRiteChannel.appendLine("Open Document fileName: " + openTextDoc.fileName);

              openTextDoc.then(srcFile => {
                iRiteChannel.show();
                iRiteChannel.appendLine("srcFile: " + srcFile.fileName);
                vscode.window.showTextDocument(srcFile).then(srcEditor => {
                  checkEngineDefined().then(engineType => {
                    checkEnginePathDefined(engineType.toString()).then(engineParameters => {
                      //got engine path and compiler from config, now init active window and push to iRiteProcessor
                      //iRiteProcessor takes argument of desired .src file path
                      let enginePath: string = vscode.workspace
                        .getConfiguration("irite")
                        .get("build.enginePath", "");
                      let compilerPath: string = vscode.workspace
                        .getConfiguration("irite")
                        .get("build.compilerPath", "");
                      iRiteChannel.appendLine("Building: " + srcFile.fileName);
                      iRiteChannel.appendLine("srcEditor: " + srcEditor.document.fileName);

                      let filepath = srcFile.fileName;
                      var path = enginePath;

                      cp.execFile(
                        path,
                        [filepath, compilerPath, "build"],
                        function (error, data, stderr) {
                          if (error != null) {
                            iRiteChannel.appendLine("Compile Function ERROR: " + error + "  stderr:" + stderr + "|");
                            console.log(error);
                            console.log(stderr);
                          }else{
                            iRiteChannel.appendLine(data);
                          }
                        }
                      );
                    })
                      .catch(error => {
                        vscode.window.setStatusBarMessage("Error 112: " + error, 5000)
                        vscode.window.showErrorMessage("Open Folder to Build Properly");
                        iRiteChannel.appendLine("Error 114:  checkEnginePathDefined: " + error);
                      });
                  })
                    .catch(error => {
                      vscode.window.showErrorMessage("Error 118: " + error);
                      vscode.window.setStatusBarMessage("Error 119:  + checkEngineDefined: " + error, 5000);
                      iRiteChannel.appendLine("Error 120: checkEngineDefined: " + error);
                    });
                },
                  err => {
                    vscode.window.showErrorMessage("Error 124: " + err);
                    vscode.window.setStatusBarMessage("Error 125: " + err, 5000);
                    iRiteChannel.appendLine("Error 126: " + err);
                  }
                );
              },
                err => {
                  vscode.window.showErrorMessage(
                    "Error 136: Invalid Workspace.  Please open folder/workspace and rebuild | "
                  );
                  iRiteChannel.appendLine(
                    "Error 137: Invalid Workspace.  Please open folder/workspace and rebuild"
                  );
                  vscode.window.setStatusBarMessage(
                    "Error 138: Invalid Workspace.  Please open folder/workspace and rebuild | ",
                    5000
                  );
                }
              );
            }
          } else {
            vscode.window.showErrorMessage(
              "Error 140: Invalid Workspace.  Please open folder/workspace and rebuild"
            );
            iRiteChannel.appendLine(
              "\n***********************************************************************************" +
                "\n********* Invalid Workspace.  Please open folder / workspace and rebuild **********" +
                "\n***********************************************************************************\n"
            );
          }


      },
      reason => {
        vscode.window.showErrorMessage(reason + "\nUnable to Save!\n\n");
        vscode.window.setStatusBarMessage(reason + "\nUnable to Save!\n\n");
        iRiteChannel.append(
          "*************************************\n" +
            "*************************************\n" +
            "Unable to Save!" +
            "*************************************\n" +
            "*************************************\n"
        );
        iRiteChannel.append("\nError 171: Please open folder(workspace).\n");
      }
    );
  });

  //registercommands takes reference from package.json, alter that file for more/different commands
  vscode.commands.registerCommand("irite.deploy", () => {
    // vscode.commands.executeCommand("irite.build");  // WOrry about this later :)
    checkEngineDefined()
      .then(engineType => {
        checkEnginePathDefined(engineType.toString())
          .then(engineParameters => {
            //got engine path and compiler from config, now init active window and push to iRiteProcessor
            //iRiteProcessor takes argument of desired .src file path
            let enginePath: string = vscode.workspace
              .getConfiguration("irite")
              .get("build.enginePath", "");
            let compilerPath: string = vscode.workspace
              .getConfiguration("irite")
              .get("build.compilerPath", "");
            let textEditor = vscode.window.activeTextEditor;
            iRiteChannel.append(
              "*************************************\n" +
                "*************************************\n" +
                "iRite Deploying: " +
                textEditor.document.fileName +
                "\n"
            );
            let filepath = textEditor.document.fileName;
            var path = enginePath;

            cp.execFile(path, [filepath, compilerPath, "deploy"], function (
              error,
              data,
              stderr
            ) {
              if (stderr != null) {
                console.log(error);
                console.log(stderr);
              }
              iRiteChannel.append(data);
            });
          })
          .catch(error => {
            vscode.window.showErrorMessage(error);
            iRiteChannel.appendLine(error);
          });
      })
      .catch(error => {
        vscode.window.setStatusBarMessage("checkEngineDefined: " + error, 5000);
        iRiteChannel.appendLine("checkEngineDefined: " + error);
      });
  });

  function checkEngineDefined() {
    return new Promise((resolve, reject) => {
      let engineType: string = vscode.workspace
        .getConfiguration("irite")
        .get("build.engine", "");
      if (engineType == "") {
        var optionRev = <vscode.MessageItem>{
          title: "revolution"
        };
        var optionTest = <vscode.MessageItem>{
          title: "Test"
        };
        vscode.window
          .showErrorMessage(
            'The "irite.build.engine" setting is not defined. Do you want to download Revolution?',
            optionRev,
            optionTest
          )
          .then(option => {
            // nothing selected
            if (typeof option == "undefined") {
              reject("undefined");
              return;
            }
            switch (option.title) {
              case optionRev.title:
                opener(
                  "https://www.ricelake.com/en-us/products/product-details/revolution-scale-software#/resources-downloads"
                );
                break;
              case optionTest.title:
                opener("http://www.ricelake.com");
                break;
              default:
                break;
            }
            reject("hyperlink");
          });
      } else {
        resolve(engineType);
      }
    });
  }

  function checkEnginePathDefined(engine: string) {
    return new Promise((resolve, reject) => {
      let enginePath: string = vscode.workspace
        .getConfiguration("irite")
        .get("build.enginePath", "");
      if (enginePath == "") {
        reject(
          'The "irite.build.enginePath" setting is not defined. Please configure.'
        );
        return;
      }
      resolve(enginePath);
    });
  }
}


src/activeEditorTracker.ts

'use strict';
import { commands, Disposable, TextEditor, window } from 'vscode';
import { BuiltInCommands } from './constants';

export class ActiveEditorTracker extends Disposable {

    private _disposable: Disposable;
    private _resolver: ((value?: TextEditor | PromiseLike<TextEditor>) => void) | undefined;

    constructor() {
        super(() => this.dispose());

        this._disposable = window.onDidChangeActiveTextEditor(e => this._resolver && this._resolver(e));
    }

    dispose() {
        this._disposable && this._disposable.dispose();
    }

    async awaitClose(timeout: number = 500): Promise<TextEditor> {
        this.close();
        return this.wait(timeout);
    }

    async awaitNext(timeout: number = 500): Promise<TextEditor> {
        this.next();
        return this.wait(timeout);
    }

    async close(): Promise<{} | undefined> {
        return commands.executeCommand(BuiltInCommands.CloseActiveEditor);
    }

    async next(): Promise<{} | undefined> {
        return commands.executeCommand(BuiltInCommands.NextEditor);
    }

    async wait(timeout: number = 500): Promise<TextEditor> {
        const editor = await new Promise<TextEditor>((resolve, reject) => {
            let timer: any;

            this._resolver = (editor: TextEditor) => {
                if (timer) {
                    clearTimeout(timer as any);
                    timer = 0;
                    resolve(editor);
                }
            };

            timer = setTimeout(() => {
                resolve(window.activeTextEditor);
                timer = 0;
            }, timeout) as any;
        });
        this._resolver = undefined;
        return editor;
    }
}


src/extension.ts

'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";
//import * as builder from './build'; //removed until language server build starts
import fs = require("fs");
import path = require("path");
import cp = require("child_process");
// import { commands, Disposable, ExtensionContext, TextEditor, window } from 'vscode';
// import { ActiveEditorTracker } from './activeEditorTracker';
// import { TextEditorComparer } from './comparers';
// import { WorkspaceState } from './constants';
// import { Logger } from './logger';
// import { ISavedEditor, SavedEditor } from './savedEditor';

var opener = require("opener");
var iRiteChannel = vscode.window.createOutputChannel("iRite Information");

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  interface EngineParams {
    engine: string;
    enginePath: string;
  }

  // Use the console to output diagnostic information (console.log) and errors (console.error)
  // This line of code will only be executed once when your extension is activated
  console.log("Congratulations, iRite is now active!");

  //check to see if the .exe is in specified path, if not display message to download revolution
  let enginePath: string = vscode.workspace
    .getConfiguration("irite")
    .get("build.enginePath", "");
  var optionRev = <vscode.MessageItem>{ title: "Get Revolution" };

  fs.stat(enginePath, function(err, stat) {
    if (err == null) {
      console.log("Revolution exists");
    } else if (err.code == "ENOENT") {
      // file does not exist
      vscode.window
        .showInformationMessage(
          "Revolution can not be located, would you like to download it?",
          optionRev
        )
        .then(option => {
          if (typeof option == "undefined") {
            return;
          }
          switch (option.title) {
            case optionRev.title:
              opener(
                "https://www.ricelake.com/en-us/products/product-details/revolution-scale-software#/resources-downloads"
              );
              break;
            default:
              break;
          }
        });
    } else {
      console.log("Revolution Checker: ", err.code);
    }
  });

  //registercommands takes reference from package.json, alter that file for more/different commands
  vscode.commands.registerCommand("irite.build", () => {
    iRiteChannel.clear();
    vscode.workspace.saveAll().then(
      success => {
        if (success) {
          iRiteChannel.appendLine("All Files Saved");

          var openTextDoc: vscode.TextDocument;

          // async function save () {
          //   try {
          //     const editorTracker = new ActiveEditorTracker();

          //     let active = window.activeTextEditor;
          //     let editor = active;
          //     const openEditors: TextEditor[] = [];
          //     do {
          //         if (editor != null) {
          //             // If we didn't start with a valid editor, set one once we find it
          //             if (active === undefined) {
          //                 active = editor;
          //             }

          //             openEditors.push(editor);
          //         }

          //         editor = await editorTracker.awaitNext(500);
          //         if (editor !== undefined && openEditors.some(_ => TextEditorComparer.equals(_, editor, { useId: true, usePosition: true }))) break;
          //     } while ((active === undefined && editor === undefined) || !TextEditorComparer.equals(active, editor, { useId: true, usePosition: true }));

          //       editorTracker.dispose();

          //       const editors = openEditors
          //           .filter(_ => _.document !== undefined)
          //           .map(_ => {
          //               return {
          //                   uri: _.document.uri,
          //                   viewColumn: _.viewColumnnp
          //               } as ISavedEditor;
          //           });

          //       this.context.workspaceState.update(WorkspaceState.SavedDocuments, editors);
          //   }
          //   catch (ex) {
          //       Logger.error(ex, 'DocumentManager.save');
          //   }
          // }


          openTextDoc = vscode.window.activeTextEditor.document;
          iRiteChannel.appendLine("Open File: " + openTextDoc.fileName);

          iRiteChannel.show();
          vscode.window.showTextDocument(openTextDoc).then(srcEditor => {
            checkEngineDefined().then(engineType => {
              checkEnginePathDefined(engineType.toString()).then(engineParameters => {
                //got engine path and compiler from config, now init active window and push to iRiteProcessor
                //iRiteProcessor takes argument of desired .src file path
                let enginePath: string = vscode.workspace
                  .getConfiguration("irite")
                  .get("build.enginePath", "");
                let compilerPath: string = vscode.workspace
                  .getConfiguration("irite")
                  .get("build.compilerPath", "");
                let filepath = openTextDoc.fileName;
                iRiteChannel.appendLine("Building: " + filepath);

                var path = enginePath;

                cp.execFile(
                  path,
                  [filepath, compilerPath, "build"],
                  function (error, data, stderr) {
                    if (error != null) {
                      iRiteChannel.appendLine("Compile Function ERROR: " + error + "  stderr:" + stderr + "|");
                      console.log(error);
                      console.log(stderr);
                    }else{
                      iRiteChannel.appendLine(data);
                    }
                  }
                );
              })
                .catch(error => {
                  vscode.window.setStatusBarMessage("Error 112: " + error, 5000)
                  vscode.window.showErrorMessage("Open Folder to Build Properly");
                  iRiteChannel.appendLine("Error 114:  checkEnginePathDefined: " + error);
                });
            })
              .catch(error => {
                vscode.window.showErrorMessage("Error 118: " + error);
                vscode.window.setStatusBarMessage("Error 119:  + checkEngineDefined: " + error, 5000);
                iRiteChannel.appendLine("Error 120: checkEngineDefined: " + error);
              });
          },
            err => {
              vscode.window.showErrorMessage("Error 124: " + err);
              vscode.window.setStatusBarMessage("Error 125: " + err, 5000);
              iRiteChannel.appendLine("Error 126: " + err);
            }
          );

        } else {
          vscode.window.setStatusBarMessage("Error Saving Files SBMsg", 5000);
          vscode.window.showErrorMessage("Error Saving Files EM");
          iRiteChannel.append(
            "\n\n*************************************\nError Saving Files.  Rectify and rebuild.\n*************************************\n\n"
          );
        }
      },
      reason => {
        vscode.window.showErrorMessage(reason + "\nUnable to Save!\n\n");
        vscode.window.setStatusBarMessage(reason + "\nUnable to Save!\n\n");
        iRiteChannel.append(
          "*************************************\n" +
            "*************************************\n" +
            "Unable to Save!" +
            "*************************************\n" +
            "*************************************\n"
        );
        iRiteChannel.append("\nError 171: Please open folder(workspace).\n");
      }
    );
  });

  //registercommands takes reference from package.json, alter that file for more/different commands
  vscode.commands.registerCommand("irite.deploy", () => {
    // vscode.commands.executeCommand("irite.build");  // WOrry about this later :)
    checkEngineDefined()
      .then(engineType => {
        checkEnginePathDefined(engineType.toString())
          .then(engineParameters => {
            //got engine path and compiler from config, now init active window and push to iRiteProcessor
            //iRiteProcessor takes argument of desired .src file path
            let enginePath: string = vscode.workspace
              .getConfiguration("irite")
              .get("build.enginePath", "");
            let compilerPath: string = vscode.workspace
              .getConfiguration("irite")
              .get("build.compilerPath", "");
            let textEditor = vscode.window.activeTextEditor;
            iRiteChannel.append(
              "*************************************\n" +
                "*************************************\n" +
                "iRite Deploying: " +
                textEditor.document.fileName +
                "\n"
            );
            let filepath = textEditor.document.fileName;
            var path = enginePath;

            cp.execFile(path, [filepath, compilerPath, "deploy"], function (
              error,
              data,
              stderr
            ) {
              if (stderr != null) {
                console.log(error);
                console.log(stderr);
              }
              iRiteChannel.append(data);
            });
          })
          .catch(error => {
            vscode.window.showErrorMessage(error);
            iRiteChannel.appendLine(error);
          });
      })
      .catch(error => {
        vscode.window.setStatusBarMessage("checkEngineDefined: " + error, 5000);
        iRiteChannel.appendLine("checkEngineDefined: " + error);
      });
  });

  function checkEngineDefined() {
    return new Promise((resolve, reject) => {
      let engineType: string = vscode.workspace
        .getConfiguration("irite")
        .get("build.engine", "");
      if (engineType == "") {
        var optionRev = <vscode.MessageItem>{
          title: "revolution"
        };
        var optionTest = <vscode.MessageItem>{
          title: "Test"
        };
        vscode.window
          .showErrorMessage(
            'The "irite.build.engine" setting is not defined. Do you want to download Revolution?',
            optionRev,
            optionTest
          )
          .then(option => {
            // nothing selected
            if (typeof option == "undefined") {
              reject("undefined");
              return;
            }
            switch (option.title) {
              case optionRev.title:
                opener(
                  "https://www.ricelake.com/en-us/products/product-details/revolution-scale-software#/resources-downloads"
                );
                break;
              case optionTest.title:
                opener("http://www.ricelake.com");
                break;
              default:
                break;
            }
            reject("hyperlink");
          });
      } else {
        resolve(engineType);
      }
    });
  }

  function checkEnginePathDefined(engine: string) {
    return new Promise((resolve, reject) => {
      let enginePath: string = vscode.workspace
        .getConfiguration("irite")
        .get("build.enginePath", "");
      if (enginePath == "") {
        reject(
          'The "irite.build.enginePath" setting is not defined. Please configure.'
        );
        return;
      }
      resolve(enginePath);
    });
  }
}